opentelemetry-sdk/src/metrics/periodic_reader_with_own_thread.rs

```diff
@@ -0,0 +1,796 @@
+        #[allow(unused_variables)]
+        if let Err(e) = result_thread_creation {
+            otel_error!(
+                name: "PeriodReaderThreadStartError",
+                event_name =  "PeriodReaderThreadStartError",
+                error = format!("{:?}", e)
+            );
+        }
+        reader
+    }
+
+    fn collect_and_export(&self, timeout: Duration) -> MetricResult<()> {
+        self.inner.collect_and_export(timeout)
+    }
+}
+
+impl fmt::Debug for PeriodicReaderWithOwnThread {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("PeriodicReaderWithOwnThread").finish()
+    }
+}
+
+struct PeriodicReaderInner {
+    exporter: Arc<dyn PushMetricExporter>,
+    message_sender: Arc<Mutex<mpsc::Sender<Message>>>,
+    producer: Mutex<Option<Weak<dyn SdkProducer>>>,
+    is_shutdown: AtomicBool,
+}
+
+impl PeriodicReaderInner {
+    fn register_pipeline(&self, producer: Weak<dyn SdkProducer>) {
+        let mut inner = self.producer.lock().expect("lock poisoned");
+        *inner = Some(producer);
+    }
+
+    fn temporality(&self, _kind: InstrumentKind) -> Temporality {
+        self.exporter.temporality()
+    }
+
+    fn collect(&self, rm: &mut ResourceMetrics) -> MetricResult<()> {
+        if self.is_shutdown.load(std::sync::atomic::Ordering::Relaxed) {
+            return Err(MetricError::Other("reader is shut down".into()));
+        }
+
+        let producer = self.producer.lock().expect("lock poisoned");
+        if let Some(p) = producer.as_ref() {
+            p.upgrade()
+                .ok_or_else(|| MetricError::Other("pipeline is dropped".into()))?
+                .produce(rm)?;
+            Ok(())
+        } else {
+            Err(MetricError::Other("pipeline is not registered".into()))
+        }
+    }
+
+    fn collect_and_export(&self, _timeout: Duration) -> MetricResult<()> {
+        // TODO: Reuse the internal vectors. Or refactor to avoid needing any
+        // owned data structures to be passed to exporters.
+        let mut rm = ResourceMetrics {
+            resource: Resource::empty(),
+            scope_metrics: Vec::new(),
+        };
+
+        let collect_result = self.collect(&mut rm);
+        #[allow(clippy::question_mark)]
+        if let Err(e) = collect_result {
+            otel_warn!(
+                name: "PeriodReaderCollectError",
+                event_name = "PeriodReaderCollectError",
+                error = format!("{:?}", e)
+            );
+            return Err(e);
+        }
+
+        if rm.scope_metrics.is_empty() {
+            otel_debug!(name: "NoMetricsCollected");
+            return Ok(());
+        }
+
+        // TODO: substract the time taken for collect from the timeout. collect
+        // involves observable callbacks too, which are user defined and can
+        // take arbitrary time.
+        //
+        // Relying on futures executor to execute async call.
+        // TODO: Add timeout and pass it to exporter.
```

Comment:
We previously discussed this approach, though I don’t fully recall if we agreed on it or if we’re open to either option:
- Passing the timeout to the exporter
- Using manual polling within a loop, checking for timeout in each iteration.
If we're still considering both options, it might be best to add a TODO comment for clarity."

Reply:
added